### PR TITLE
Allow to only check UDP ports

### PR DESCRIPTION
**Pull request by ydkn**
#### General

- [x] Update Changelog following the conventions laid out [here](https://github.com/sensu-plugins/community/blob/master/HOW_WE_CHANGELOG.md)

- [x] Update README with any necessary configuration snippets

- [x] Binstubs are created if needed

- [x] RuboCop passes

- [x] Existing tests pass

#### Purpose

The current _check-ports-bind.rb_ check does not allow to perform a check only via UDP.
This PR extends the supported syntax for specifying ports to also set the desired protocol (TCP or UDP):
address:port/proto

In my opinion the "--hard" option is no longer required due to the fact that the protocol can now specified directly
through the "--portbinds" option.
But I'm also open for suggestions and will readd the "--hard" option if required.

#### Known Compatibility Issues

Dropped the "--hard" option in favour of specifying the protocol on a per port basis (see above for explanation).

### CODE REVIEW

**Comment by majormoses** [comment]
Thanks for your contribution to Sensu plugins! Without people like you submitting PRs we couldn't run the project. I will review it shortly.

File: bin/check-ports-bind.rb [hunk 1]
```diff
@@ -105,53 +96,101 @@ def severity(warn, text)
...
+    config[:portbinds].split(',').each do |portbind|
+      portbind = "#{config[:host]}:#{portbind}" unless portbind.include?(':')
+      portbind = "#{portbind}/tcp" unless portbind.include?('/')
+
+      protocol     = portbind.split('/')[1] || 'tcp'
+      address_port = portbind.split('/')[0]
+      address      = address_port.split(':')[0]
+      port         = address_port.split(':')[1].to_i
+
+      if port =~ /^[0-9]+(-[0-9]+)$/
```
Comment (majormoses) [inline bin/check-ports-bind.rb]: we should also check that they are actually between 0-65k while we are at it. I'd vote for a function called `valid_port?(port)`

File: bin/check-ports-bind.rb [hunk 2]
```diff
@@ -105,53 +96,101 @@ def severity(warn, text)
...
+        # Single port
+
+        binds += portbind_hashs(address, port, protocol)
+      end
+    end
+
+    binds
+  end
+
+  def portbind_hashs(address, port, protocol)
```
Comment (majormoses) [inline bin/check-ports-bind.rb]: this returns an array of hashes, perhaps just `portbindings`?

File: bin/check-ports-bind.rb (hunk 1 above)
Comment (ydkn) [reply bin/check-ports-bind.rb]: Great idea. What do you think about returning a critical result if this validation fails?

**Comment by ydkn** [comment]
We could drop the _Impact: Breaking Change_ label since it is now backward compatible.

**Comment by jspaleta** [comment]
Hey @ydkn  and @majormoses 

It appears that there are still some unresolved comments from the earlier discussion.
Is this ready to merge?

I just did a quick conflict merge resolution on the changelog to clear that minor conflict.

**Comment by majormoses** [comment]
I think my only ask is still to check that its actually a valid port range rather than regex.

**Comment by ydkn** [comment]
I've added a check for port numbers and port ranges.

**Comment by majormoses** [comment]
Working on cutting a release, had to re-read the changes looks like the ruby versions were dropped in another release so this can go out without breaking changes.

**Comment by majormoses** [comment]
I lied there are other unrelated breaking changes that are not part of this PR so it will go out as 5.0.0

**Comment by majormoses** [comment]
A release has been cut and we should see builds be published as soon as CI and CD have run their course.

**Comment by majormoses** [comment]
released: https://rubygems.org/gems/sensu-plugins-network-checks/versions/5.0.0

**Comment by majormoses** [comment]
Sorry for the long back and forth, keep up the great work.

**Pull request merged**